lib/smack/Prelude.cpp

```diff
@@ -280,42 +423,228 @@ std::string getBytewisePointerStorageStr() {
   return s.str();
 }
 
-std::string Prelude::getPrelude() {
-  const std::vector<unsigned> INTEGER_SIZES {
-    1, 5, 6, 8, 16, 24, 32, 40, 48, 56, 64, 80, 88, 96, 128, 160, 256
-  };
+const std::vector<unsigned> INTEGER_SIZES {
+  1, 5, 6, 8, 16, 24, 32, 40, 48, 56, 64, 80, 88, 96, 128, 160, 256
+};
 
-  const std::vector<unsigned> REF_CONSTANTS {
-    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 1024
-  };
+//const std::vector<unsigned> REF_CONSTANTS {
+//  0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 1024
+//};
 
-  const std::vector<std::string> BINARY_OPS {
-    "add", "sub", "mul", "sdiv", "smod", "srem", "udiv",
-    "urem", "shl", "lshr", "ashr", "and", "or", "xor", "nand"
-  };
+template <BinExpr::Binary OP>
+const Expr* intArithExpr(unsigned size) {
+  return new BinExpr(OP, makeIntVarExpr(1), makeIntVarExpr(2));
+}
 
-  const std::vector<std::string> COMP_OPS {
-    "ule", "ult", "uge", "ugt", "sle", "slt", "sge", "sgt"
-  };
+template <bool SIGN, bool MIN>
+const Expr* bvMinMaxExpr(unsigned size) {
+  std::string signLetter = SIGN? "s" : "u";
+  std::string pred = MIN? "lt" : "gt";
+  return Expr::if_then_else(
+    Expr::fn(indexedName("$"+signLetter+pred,
+      {getBvTypeName(size), Naming::BOOL_TYPE}),
+      makeIntVarExprs(2)),
+    makeIntVarExpr(1), makeIntVarExpr(2));
+}
 
-  const std::vector<std::string> EQ_OPS {"eq", "ne"};
+template <bool MIN>
+const Expr* intMinMaxExpr(unsigned size) {
+  const Expr* a1 = makeIntVarExpr(1);
+  const Expr* a2 = makeIntVarExpr(2);
+  auto pred = MIN? Expr::lt(a1, a2) : Expr::lt(a2, a1);
+  return Expr::if_then_else(pred, a1, a2);
+}
 
-  const std::vector<BinExpr::Binary> BINARY_OPS_BODY {
-    BinExpr::Plus, BinExpr::Minus, BinExpr::Times
+void Prelude::generateIntArithOps(std::stringstream& s) {
+  typedef const Expr* (*bodyT)(unsigned size);
```

Comment:
For each "block" of axioms or functions that we generate in the prelude, it would be good to write a bpl comment into the boogie file saying, for example, `// Models for integer arithmetic operations`. This would allow us to more easily track down which function in `Prelude.cpp` generated which parts of the Boogie prelude source. I hope this makes sense.